packages/db/src/lib/duplicate-detection.ts

```diff
@@ -0,0 +1,92 @@
+import { Knex } from 'knex';
+import { deconstructCollectionId } from '@cumulus/message/Collections';
+import { CollectionPgModel } from '../models/collection';
+
+type GranuleInput = {
+  collectionId: string;
+  producerGranuleId: string;
+  collectionCumulusId?: number;
+  granuleId?: string;
+};
+
+type GranuleRecord = {
+  cumulus_id: number;
+  granule_id: string;
+  producer_granule_id: string;
+  collection_cumulus_id: number;
+  status: string;
+  group_state: string | null;
+};
+
+type DuplicateGranulesResult = {
+  sameCollectionMatches: GranuleRecord[];
+  differentCollectionMatches: GranuleRecord[];
+  customCriteriaMatches: GranuleRecord[];
+};
+
+export const getNextGranuleGroupId = async (knex: Knex) : Promise<number> => {
+  const result = await knex.raw("SELECT nextval('granule_group_id_seq')::int");
+  return result.rows[0].nextval;
+};
+
+/**
+ * Detects "active" duplicate granules based on producerGranuleId and collection.
+ * A granule is considered "active" if its `granule_groups.state != 'H'` or it's
+ * not in granule_groups.
+ */
+export const findDuplicateGranules = async (
+  incoming: GranuleInput,
+  knex: Knex,
+  collectionPgModel = new CollectionPgModel()
+): Promise<DuplicateGranulesResult> => {
+  const {
+    collectionId,
+    producerGranuleId,
+    granuleId,
+  } = incoming;
+
+  let { collectionCumulusId } = incoming;
+
+  // Ensure we have collectionCumulusId
+  if (!collectionCumulusId) {
+    const pgCollection = await collectionPgModel.get(
+      knex,
+      deconstructCollectionId(collectionId)
+    );
+    collectionCumulusId = pgCollection.cumulus_id;
+  }
+
+  // Query for active granules with matching producerGranuleId
+  const duplicates = await knex('granules')
+    .leftJoin('granule_groups', 'granules.cumulus_id', 'granule_groups.granule_cumulus_id')
+    .where('granules.producer_granule_id', producerGranuleId)
+    .modify((qb) => {
+      if (granuleId) {
+        qb.whereNot('granules.granule_id', granuleId);
```

Comment:
I'm assuming from the tests the intent here is to allow providing a single exclusion granuleId if you're looking for all other records for <granuleId> versus just a group?    That seems like a good thing to have!      See comments at top of function re: docstring. 

Nit: Is there ever a reasonable situation where we'd want that to be a list?